log-manager/src/main/java/io/airlift/log/SocketMessageOutput.java

```diff
@@ -0,0 +1,99 @@
+
+import com.google.common.net.HostAndPort;
+import org.weakref.jmx.Managed;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Formatter;
+
+public class SocketMessageOutput
+        implements MessageOutput
+{
+    private static final int CONNECTION_TIMEOUT_MILLIS = 100;
+    private static final int MAX_WRITE_ATTEMPTS_PER_MESSAGE = 5;
+
+    private final InetSocketAddress socketAddress;
+    private final AtomicLong failedConnections = new AtomicLong(0);
+
+    @GuardedBy("this")
+    private Socket socket;
+    @GuardedBy("this")
+    private OutputStream currentOutputStream;
+
+    private SocketMessageOutput(String hostAndPortString)
+    {
+        HostAndPort hostAndPort = HostAndPort.fromString(hostAndPortString);
+        this.socketAddress = new InetSocketAddress(hostAndPort.getHost(), hostAndPort.getPort());
+    }
+
+    public static BufferedHandler createSocketHandler(String logPath, Formatter formattter)
+    {
+        SocketMessageOutput output = new SocketMessageOutput(logPath);
+        BufferedHandler handler = new BufferedHandler(output, formattter);
+        handler.start();
+        return handler;
+    }
+
+    @Override
+    public synchronized void writeMessage(byte[] message)
+            throws IOException
+    {
+        IOException exception = new IOException(String.format("Exception thrown attempting to write log messages"));
+        for (int i = 0; i < MAX_WRITE_ATTEMPTS_PER_MESSAGE; i++) {
+            if (socket == null || socket.isClosed()) {
+                try {
+                    socket = new Socket();
+                    socket.connect(socketAddress, CONNECTION_TIMEOUT_MILLIS);
+                    currentOutputStream = socket.getOutputStream();
+                }
+                catch (IOException e) {
+                    failedConnections.incrementAndGet();
+                    exception.addSuppressed(new IOException("Error connecting via socket to " + socketAddress.getHostName() + " on port " + socketAddress.getPort(), e));
+                }
+            }
+            try {
+                currentOutputStream.write(message);
+                break;
+            }
+            catch (IOException e) {
+                socket = null;
+                exception.addSuppressed(e);
+            }
+        }
+        if (exception.getSuppressed().length > 0) {
+            throw exception;
+        }
+    }
+
+    @Override
+    public synchronized void flush()
+            throws IOException
+    {
+        if (currentOutputStream != null) {
+            currentOutputStream.flush();
+        }
+    }
+
+    @Override
+    public synchronized void close()
+            throws IOException
+    {
+        currentOutputStream.flush();
+        currentOutputStream.close();
+        socket.close();
```

Comment:
any of these can throw, you should collect the exceptions together using suppression, and make sure everything is nulled.